File: go/scanner/scanner.go
// Package scanner provides AST-based static analysis to detect cost-generating
// patterns in Go source files.  Uses only stdlib go/ast and go/parser — no
// external dependencies.
package scanner

import (
	"fmt"
	"go/ast"
	"go/parser"
	"go/token"
	"os"
	"path/filepath"
	"strings"
)

// CostPoint represents a detected cost-generating call site in source code.
type CostPoint struct {
	File              string `json:"file"`
	Line              int    `json:"line"`
	Category          string `json:"category"` // "llm", "http", "service", "framework", "embedding", "speech", "image"
	Provider          string `json:"provider"`
	Description       string `json:"description"`
	AutoInstrumented  bool   `json:"auto_instrumented"`
	ImportPath        string `json:"import_path"`
}

// ScanResult holds the aggregated output of a directory scan.
type ScanResult struct {
	CostPoints   []CostPoint `json:"cost_points"`
	FilesScanned int         `json:"files_scanned"`
}

// AutoCount returns the number of auto-instrumented cost points.
func (r ScanResult) AutoCount() int {
	n := 0
	for _, cp := range r.CostPoints {
		if cp.AutoInstrumented {
			n++
		}
	}
	return n
}

// ManualCount returns the number of cost points requiring manual tracking.
func (r ScanResult) ManualCount() int {
	n := 0
	for _, cp := range r.CostPoints {
		if !cp.AutoInstrumented {
			n++
		}
	}
	return n
}

// ── Skip directories ────────────────────────────────────────────────

var skipDirs = map[string]bool{
	"vendor": true, ".git": true, "node_modules": true,
	"testdata": true, ".cache": true, "dist": true, "build": true,
}

// ── Import → provider mapping ───────────────────────────────────────

// importPattern maps an import path prefix to a provider name and whether
// dexcost auto-instruments calls through this import.
type importPattern struct {
	prefix          string
	provider        string
	autoInstrumented bool
}

var knownImports = []importPattern{
	// LLM providers
	{"github.com/sashabaranov/go-openai", "openai", false},
	{"github.com/openai/openai-go", "openai", false},
	{"github.com/anthropics/anthropic-sdk-go", "anthropic", false},
	{"github.com/google/generative-ai-go", "google-ai", false},
	{"google.golang.org/genai", "google-ai", false},
	{"github.com/cohere-ai/cohere-go", "cohere", false},
	{"github.com/mistralai/client-go", "mistral", false},
	{"github.com/replicate/replicate-go", "replicate", false},
	{"github.com/ollama/ollama", "ollama", false},

	// Missing LLM providers
	{"github.com/groq/groq-go", "groq", false},
	{"github.com/deepseek-ai/deepseek-go", "deepseek", false},
	{"github.com/together-ai/together-go", "together", false},

	// Agent frameworks
	{"github.com/tmc/langchaingo", "langchaingo", false},
	{"github.com/cloudwego/eino", "eino", false},
	{"github.com/mozilla-ai/any-llm-go", "anyllm", false},
	{"github.com/aiochan/crewai-go", "crewai", false},

	// Bedrock-specific (not just generic AWS) — must come before generic AWS
	{"github.com/aws/aws-sdk-go-v2/service/bedrockruntime", "aws-bedrock", false},

	// Cloud / Infrastructure
	{"github.com/aws/aws-sdk-go-v2", "aws", false},
	{"github.com/aws/aws-sdk-go", "aws", false},
	{"cloud.google.com/go", "gcp", false},
	{"github.com/Azure/azure-sdk-for-go", "azure", false},

	// Databases
	{"go.mongodb.org/mongo-driver", "mongodb", false},
	{"github.com/elastic/go-elasticsearch", "elasticsearch", false},
	{"github.com/redis/go-redis", "redis", false},
	{"github.com/go-redis/redis", "redis", false},

	// Vector databases
	{"github.com/pinecone-io/go-pinecone", "pinecone", false},
	{"github.com/weaviate/weaviate-go-client", "weaviate", false},
	{"github.com/milvus-io/milvus-sdk-go", "milvus", false},
	{"github.com/qdrant/go-client", "qdrant", false},

	// Payments
	{"github.com/stripe/stripe-go", "stripe", false},

	// Messaging
	{"github.com/twilio/twilio-go", "twilio", false},
	{"github.com/sendgrid/sendgrid-go", "sendgrid", false},
	{"github.com/resend/resend-go", "resend", false},
	{"github.com/slack-go/slack", "slack", false},

	// Geo / Maps
	{"googlemaps.github.io/maps", "google-maps", false},

	// Scraping
	{"github.com/gocolly/colly", "colly", false},

	// HTTP clients (need manual cost tracking)
	{"net/http", "net/http", false},
}

// ── Call patterns ───────────────────────────────────────────────────

// callPattern matches a method call on an object whose import is known.
type callPattern struct {
	receiverPkg string // import short name to match (e.g. "openai")
	methodChain string // dot-separated method chain to look for
	category    string
	provider    string
	description string
	auto        bool
}

var callPatterns = []callPattern{
	// OpenAI
	{"openai", "CreateChatCompletion", "llm", "openai", "OpenAI chat completion", false},
	{"openai", "CreateCompletion", "llm", "openai", "OpenAI completion", false},
	{"openai", "CreateEmbeddings", "embedding", "openai", "OpenAI embedding", false},
	{"openai", "CreateImage", "image", "openai", "OpenAI DALL-E image", false},
	{"openai", "CreateTranscription", "speech", "openai", "OpenAI Whisper transcription", false},
	{"openai", "CreateSpeech", "speech", "openai", "OpenAI TTS", false},

	// Anthropic
	{"anthropic", "Messages", "llm", "anthropic", "Anthropic message", false},
	{"anthropic", "Complete", "llm", "anthropic", "Anthropic completion", false},

	// Google AI
	{"genai", "GenerateContent", "llm", "google-ai", "Google AI generate content", false},

	// Cohere
	{"cohere", "Chat", "llm", "cohere", "Cohere chat", false},
	{"cohere", "Generate", "llm", "cohere", "Cohere generate", false},
	{"cohere", "Embed", "embedding", "cohere", "Cohere embedding", false},

	// Replicate
	{"replicate", "Run", "llm", "replicate", "Replicate model run", false},

	// LangChainGo
	{"llms", "Call", "framework", "langchaingo", "LangChainGo LLM call", false},
	{"llms", "GenerateContent", "framework", "langchaingo", "LangChainGo generate content", false},
	{"chains", "Call", "framework", "langchaingo", "LangChainGo chain call", false},
	{"chains", "Run", "framework", "langchaingo", "LangChainGo chain run", false},
	{"agents", "Run", "framework", "langchaingo", "LangChainGo agent run", false},

	// MongoDB
	{"mongo", "Find", "service", "mongodb", "MongoDB find", false},
	{"mongo", "FindOne", "service", "mongodb", "MongoDB findOne", false},
	{"mongo", "InsertOne", "service", "mongodb", "MongoDB insertOne", false},
	{"mongo", "InsertMany", "service", "mongodb", "MongoDB insertMany", false},
	{"mongo", "UpdateOne", "service", "mongodb", "MongoDB updateOne", false},
	{"mongo", "DeleteOne", "service", "mongodb", "MongoDB deleteOne", false},
	{"mongo", "Aggregate", "service", "mongodb", "MongoDB aggregate", false},

	// Elasticsearch
	{"elasticsearch", "Search", "service", "elasticsearch", "Elasticsearch search", false},
	{"esapi", "Search", "service", "elasticsearch", "Elasticsearch search", false},

	// Redis
	{"redis", "Get", "service", "redis", "Redis GET", false},
	{"redis", "Set", "service", "redis", "Redis SET", false},
	{"redis", "Del", "service", "redis", "Redis DEL", false},

	// Stripe
	{"stripe", "New", "service", "stripe", "Stripe API call", false},
	{"paymentintent", "New", "service", "stripe", "Stripe PaymentIntent", false},
	{"charge", "New", "service", "stripe", "Stripe Charge", false},

	// Twilio
	{"twilio", "CreateMessage", "service", "twilio", "Twilio message", false},

	// Bedrock-specific
	{"bedrockruntime", "InvokeModel", "llm", "aws-bedrock", "AWS Bedrock invoke model", false},
	{"bedrockruntime", "InvokeModelWithResponseStream", "llm", "aws-bedrock", "AWS Bedrock streaming invoke", false},

	// Groq
	{"groq", "CreateChatCompletion", "llm", "groq", "Groq chat completion", false},

	// DeepSeek
	{"deepseek", "CreateChatCompletion", "llm", "deepseek", "DeepSeek chat completion", false},

	// HTTP calls
	{"http", "Get", "http", "net/http", "HTTP GET", false},
	{"http", "Post", "http", "net/http", "HTTP POST", false},
	{"http", "Do", "http", "net/http", "HTTP request", false},
}

// ── Scanner engine ──────────────────────────────────────────────────

// ScanDirectory walks a directory tree and scans all .go files for cost points.
func ScanDirectory(root string) ScanResult {
	result := ScanResult{}
	info, err := os.Stat(root)
	if err != nil {
		return result
	}
	if !info.IsDir() {
		root = filepath.Dir(root)
	}

	_ = filepath.Walk(root, func(path string, info os.FileInfo, err error) error {
		if err != nil {
			return nil
		}
		if info.IsDir() {
			if skipDirs[info.Name()] {
				return filepath.SkipDir
			}
			return nil
		}
		if !strings.HasSuffix(path, ".go") || strings.HasSuffix(path, "_test.go") {
			return nil
		}

		result.FilesScanned++
		points := scanFile(path)
		result.CostPoints = append(result.CostPoints, points...)
		return nil
	})

	return result
}

// scanFile parses a single Go file and extracts cost points using AST analysis.
func scanFile(path string) []CostPoint {
	fset := token.NewFileSet()
	node, err := parser.ParseFile(fset, path, nil, parser.ImportsOnly|parser.ParseComments)
	if err != nil {
		// Try full parse for call detection
		node, err = parser.ParseFile(fset, path, nil, 0)
		if err != nil {
			return nil
		}
	} else {
		// Re-parse with full AST for call detection
		node, err = parser.ParseFile(fset, path, nil, 0)
		if err != nil {
			return nil
		}
	}

	var points []CostPoint
	seen := make(map[int]bool) // deduplicate by line number

	// Step 1: Collect imports and map short names to full paths.
	importMap := make(map[string]string) // shortName → full import path
	for _, imp := range node.Imports {
		importPath := strings.Trim(imp.Path.Value, `"`)
		shortName := filepath.Base(importPath)
		if imp.Name != nil {
			shortName = imp.Name.Name
		}
		// Handle hyphenated package names (e.g. "go-openai" → "openai" in code)
		shortName = strings.TrimPrefix(shortName, "go-")
		importMap[shortName] = importPath
	}

	// Step 2: Detect import-level cost points (known SDKs).
	for _, imp := range node.Imports {
		importPath := strings.Trim(imp.Path.Value, `"`)
		for _, kp := range knownImports {
			if strings.HasPrefix(importPath, kp.prefix) {
				line := fset.Position(imp.Pos()).Line
				if !seen[line] {
					seen[line] = true
					points = append(points, CostPoint{
						File:             path,
						Line:             line,
						Category:         "import",
						Provider:         kp.provider,
						Description:      fmt.Sprintf("Import: %s", importPath),
						AutoInstrumented: kp.autoInstrumented,
						ImportPath:       importPath,
					})
				}
				break
			}
		}
	}

	// Step 3: Walk AST for function/method calls matching known patterns.
	ast.Inspect(node, func(n ast.Node) bool {
		call, ok := n.(*ast.CallExpr)
		if !ok {
			return true
		}

		callStr := resolveCallExpr(call.Fun)
		if callStr == "" {
			return true
		}

		for _, pat := range callPatterns {
			// Match: either "pkg.Method" or "receiver.Method" where receiver
			// is from a known import.
			parts := strings.Split(callStr, ".")
			if len(parts) < 2 {
				continue
			}
			method := parts[len(parts)-1]
			pkgOrVar := parts[0]

			if method == pat.methodChain && pkgOrVar == pat.receiverPkg {
				line := fset.Position(call.Pos()).Line
				if !seen[line] {
					seen[line] = true
					points = append(points, CostPoint{
						File:             path,
						Line:             line,
						Category:         pat.category,
						Provider:         pat.provider,
						Description:      pat.description,
						AutoInstrumented: pat.auto,
						ImportPath:       importMap[pkgOrVar],
					})
				}
				return true
			}

			// Also match deeper chains like client.Chat.Completions.Create
			if method == pat.methodChain {
				// Check if any part of the chain matches the receiver package
				for _, part := range parts[:len(parts)-1] {
					if strings.EqualFold(part, pat.receiverPkg) {
						line := fset.Position(call.Pos()).Line
						if !seen[line] {
							seen[line] = true
							points = append(points, CostPoint{
								File:             path,
								Line:             line,
								Category:         pat.category,
								Provider:         pat.provider,
								Description:      pat.description,
								AutoInstrumented: pat.auto,
								ImportPath:       importMap[pkgOrVar],
							})
						}
						return true
					}
				}
			}
		}

		return true
	})

	return points
}

// resolveCallExpr attempts to resolve a call expression to a dotted string.
// e.g., client.Chat.Completions.Create → "client.Chat.Completions.Create"
func resolveCallExpr(expr ast.Expr) string {
	switch e := expr.(type) {
	case *ast.SelectorExpr:
		prefix := resolveCallExpr(e.X)
		if prefix != "" {
			return prefix + "." + e.Sel.Name
		}
		return e.Sel.Name
	case *ast.Ident:
		return e.Name
	case *ast.CallExpr:
		// e.g., client.Method().Chain() — resolve inner call
		return resolveCallExpr(e.Fun)
	case *ast.IndexExpr:
		// e.g., generic type instantiation
		return resolveCallExpr(e.X)
	default:
		return ""
	}
}

// GenerateStubs produces record_cost snippets for manual cost points.
func GenerateStubs(result ScanResult) string {
	var sb strings.Builder

	auto := []CostPoint{}
	manual := []CostPoint{}
	for _, cp := range result.CostPoints {
		if cp.Category == "import" {
			continue
		}
		if cp.AutoInstrumented {
			auto = append(auto, cp)
		} else {
			manual = append(manual, cp)
		}
	}

	sb.WriteString("// ============================================================\n")
	sb.WriteString("// dexcost integration stubs\n")
	sb.WriteString("// Generated by: dexcost scan --generate-stubs\n")
	sb.WriteString("// ============================================================\n\n")

	sb.WriteString("// --- Step 1: Initialize dexcost ---\n")
	sb.WriteString("import (\n")
	sb.WriteString("    \"context\"\n\n")
	sb.WriteString("    dexcost \"github.com/DexwoxBusiness/dexcost-go\"\n")
	if len(manual) > 0 {
		sb.WriteString("    \"github.com/shopspring/decimal\"\n")
	}
	sb.WriteString(")\n\n")
	sb.WriteString("dexcost.Init(dexcost.Config{\n")
	sb.WriteString("    APIKey: \"dx_live_...\", // or set DEXCOST_API_KEY env var\n")
	sb.WriteString("})\n\n")

	sb.WriteString("// --- Step 2: Set customer context (in your request handler) ---\n")
	sb.WriteString("ctx := dexcost.SetContext(context.Background(), \"your_customer_id\", \"your_project_id\")\n\n")

	sb.WriteString("// --- Step 3: Track tasks ---\n")
	sb.WriteString("ctx, task := dexcost.StartTask(ctx, \"your_task_type\")\n")
	sb.WriteString("defer task.End(dexcost.StatusSuccess)\n\n")

	if len(manual) > 0 {
		sb.WriteString("// --- Manual cost tracking (for services not auto-instrumented) ---\n")
		for _, cp := range manual {
			fmt.Fprintf(&sb, "// %s:%d — %s\n", cp.File, cp.Line, cp.Description)
			fmt.Fprintf(&sb, "task.RecordCost(\"%s\", decimal.RequireFromString(\"0.00\")) // TODO: set actual cost\n\n", cp.Provider)
		}
	}

	if len(auto) > 0 {
		sb.WriteString("// --- Auto-instrumented (no code changes needed) ---\n")
		providers := map[string]int{}
		for _, cp := range auto {
			providers[cp.Provider]++
		}
		for provider, count := range providers {
			fmt.Fprintf(&sb, "// ✓ %s (%d call%s detected)\n", provider, count, pluralS(count))
		}
	}

	return sb.String()
}

func pluralS(n int) string {
	if n == 1 {
		return ""
	}
	return "s"
}

File: typescript/src/cli/scanner.ts
/**
 * Code scanner for cost point detection (US-031).
 *
 * Regex-based static analysis — no API key needed, runs offline.
 *
 * Detection layers:
 * 1. LLM SDK calls (auto-instrumentable by dexcost or manual)
 * 2. Agent framework invocations (LangChain, Mastra, CrewAI, etc.)
 * 3. Direct HTTP calls to known AI API endpoints
 * 4. SDK imports for paid services (AWS, Stripe, Firebase, etc.)
 * 5. Service-specific method calls (MongoDB, Elasticsearch, etc.)
 */

export interface CostPoint {
  file: string;
  line: number;
  provider: string;
  type:
    | "llm_call"
    | "http_call"
    | "sdk_import"
    | "framework_call"
    | "service_call"
    | "embedding"
    | "speech"
    | "image";
  pattern: string;
  autoInstrumentable: boolean;
}

interface ScanPattern {
  regex: RegExp;
  provider: string;
  type: CostPoint["type"];
  autoInstrumentable: boolean;
}

const PATTERNS: ScanPattern[] = [
  // ── LLM SDK calls (auto-instrumented by dexcost) ────────────────────
  {
    regex: /\.chat\s*\.\s*completions\s*\.\s*create\b/g,
    provider: "openai",
    type: "llm_call",
    autoInstrumentable: true,
  },
  {
    regex: /\.messages\s*\.\s*create\b/g,
    provider: "anthropic",
    type: "llm_call",
    autoInstrumentable: true,
  },
  {
    regex: /\b(generateText|streamText)\s*\(/g,
    provider: "vercel-ai",
    type: "llm_call",
    autoInstrumentable: true,
  },

  // ── Additional LLM providers (need recordCost) ──────────────────────
  // Groq — uses OpenAI-compatible SDK: groq-sdk
  {
    regex: /from\s+["']groq-sdk["']|require\s*\(\s*["']groq-sdk["']\s*\)/g,
    provider: "groq",
    type: "sdk_import",
    autoInstrumentable: false,
  },
  // Mistral — @mistralai/mistralai, uses client.chat.complete()
  {
    regex: /\.chat\s*\.\s*complete\b/g,
    provider: "mistral",
    type: "llm_call",
    autoInstrumentable: false,
  },
  {
    regex: /from\s+["']@mistralai\/mistralai["']/g,
    provider: "mistral",
    type: "sdk_import",
    autoInstrumentable: false,
  },
  // Together AI — together-ai
  {
    regex: /from\s+["']together-ai["']|require\s*\(\s*["']together-ai["']\s*\)/g,
    provider: "together",
    type: "sdk_import",
    autoInstrumentable: false,
  },
  // Replicate — replicate.run()
  {
    regex: /replicate\s*\.\s*run\s*\(/g,
    provider: "replicate",
    type: "llm_call",
    autoInstrumentable: false,
  },
  {
    regex: /from\s+["']replicate["']|require\s*\(\s*["']replicate["']\s*\)/g,
    provider: "replicate",
    type: "sdk_import",
    autoInstrumentable: false,
  },
  // Fireworks AI
  {
    regex: /from\s+["']@fireworks-ai\/client["']/g,
    provider: "fireworks",
    type: "sdk_import",
    autoInstrumentable: false,
  },
  // Cohere — cohere.chat() / cohere.generate()
  {
    regex: /\bcohere\s*\.\s*(chat|generate|embed)\s*\(/g,
    provider: "cohere",
    type: "llm_call",
    autoInstrumentable: true,
  },
  {
    regex: /from\s+["']cohere-ai["']|require\s*\(\s*["']cohere-ai["']\s*\)/g,
    provider: "cohere",
    type: "sdk_import",
    autoInstrumentable: true,
  },
  // Google Gemini — @google/genai, generateContent
  {
    regex: /\.generateContent\s*\(/g,
    provider: "google-ai",
    type: "llm_call",
    autoInstrumentable: true,
  },
  // Ollama
  {
    regex: /from\s+["']ollama["']|require\s*\(\s*["']ollama["']\s*\)/g,
    provider: "ollama",
    type: "sdk_import",
    autoInstrumentable: false,
  },

  // ── Embeddings ──────────────────────────────────────────────────────
  {
    regex: /\.embeddings\s*\.\s*create\s*\(/g,
    provider: "openai",
    type: "embedding",
    autoInstrumentable: false,
  },

  // ── Speech / Audio ──────────────────────────────────────────────────
  {
    regex: /\.audio\s*\.\s*transcriptions\s*\.\s*create\s*\(/g,
    provider: "openai-whisper",
    type: "speech",
    autoInstrumentable: false,
  },
  {
    regex: /\.audio\s*\.\s*speech\s*\.\s*create\s*\(/g,
    provider: "openai-tts",
    type: "speech",
    autoInstrumentable: false,
  },
  {
    regex: /from\s+["']@deepgram\/sdk["']/g,
    provider: "deepgram",
    type: "sdk_import",
    autoInstrumentable: false,
  },
  {
    regex: /from\s+["']assemblyai["']/g,
    provider: "assemblyai",
    type: "sdk_import",
    autoInstrumentable: false,
  },

  // ── Image generation ────────────────────────────────────────────────
  {
    regex: /\.images\s*\.\s*generate\s*\(/g,
    provider: "openai-dalle",
    type: "image",
    autoInstrumentable: false,
  },

  // ── Agent framework calls ───────────────────────────────────────────
  // LangChain.js — .invoke(), .stream(), .batch()
  {
    regex: /from\s+["'](@langchain\/openai|@langchain\/anthropic|@langchain\/core|langchain)["']/g,
    provider: "langchain",
    type: "sdk_import",
    autoInstrumentable: false,
  },
  {
    regex: /\.(invoke|ainvoke|stream|batch)\s*\(/g,
    provider: "langchain",
    type: "framework_call",
    autoInstrumentable: false,
  },
  // LangGraph
  {
    regex: /from\s+["']@langchain\/langgraph["']/g,
    provider: "langgraph",
    type: "sdk_import",
    autoInstrumentable: false,
  },
  // Mastra
  {
    regex: /from\s+["']@mastra\/core["']/g,
    provider: "mastra",
    type: "sdk_import",
    autoInstrumentable: false,
  },
  // CrewAI (JS)
  {
    regex: /from\s+["']crewai["']/g,
    provider: "crewai",
    type: "sdk_import",
    autoInstrumentable: false,
  },

  // ── Direct HTTP calls to known AI API endpoints ─────────────────────
  {
    regex: /fetch\s*\(\s*["'`]https?:\/\/api\.openai\.com/g,
    provider: "openai",
    type: "http_call",
    autoInstrumentable: false,
  },
  {
    regex: /fetch\s*\(\s*["'`]https?:\/\/api\.anthropic\.com/g,
    provider: "anthropic",
    type: "http_call",
    autoInstrumentable: false,
  },
  {
    regex: /fetch\s*\(\s*["'`]https?:\/\/api\.cohere\.(ai|com)/g,
    provider: "cohere",
    type: "http_call",
    autoInstrumentable: false,
  },
  {
    regex: /fetch\s*\(\s*["'`]https?:\/\/api\.groq\.com/g,
    provider: "groq",
    type: "http_call",
    autoInstrumentable: false,
  },
  {
    regex: /fetch\s*\(\s*["'`]https?:\/\/api\.mistral\.ai/g,
    provider: "mistral",
    type: "http_call",
    autoInstrumentable: false,
  },
  {
    regex: /fetch\s*\(\s*["'`]https?:\/\/api\.together\.xyz/g,
    provider: "together",
    type: "http_call",
    autoInstrumentable: false,
  },
  {
    regex: /fetch\s*\(\s*["'`]https?:\/\/api\.replicate\.com/g,
    provider: "replicate",
    type: "http_call",
    autoInstrumentable: false,
  },
  {
    regex: /fetch\s*\(\s*["'`]https?:\/\/api\.deepseek\.com/g,
    provider: "deepseek",
    type: "http_call",
    autoInstrumentable: false,
  },

  // ── Cloud / Infrastructure SDK imports ──────────────────────────────
  {
    regex: /@aws-sdk\/client-bedrock-runtime/g,
    provider: "aws-bedrock",
    type: "sdk_import",
    autoInstrumentable: true,
  },
  {
    regex: /@aws-sdk\/client-/g,
    provider: "aws",
    type: "sdk_import",
    autoInstrumentable: false,
  },
  {
    regex: /@google-ai\/generativelanguage|@google\/generative-ai|@google\/genai/g,
    provider: "google-ai",
    type: "sdk_import",
    autoInstrumentable: true,
  },
  {
    regex: /from\s+["']@google-cloud\//g,
    provider: "gcp",
    type: "sdk_import",
    autoInstrumentable: false,
  },

  // ── Payments ────────────────────────────────────────────────────────
  {
    regex: /from\s+["']stripe["']|require\s*\(\s*["']stripe["']\s*\)/g,
    provider: "stripe",
    type: "sdk_import",
    autoInstrumentable: false,
  },
  {
    regex: /stripe\s*\.\s*(paymentIntents|charges|customers|subscriptions|invoices)\s*\.\s*(create|retrieve|update|list)\s*\(/g,
    provider: "stripe",
    type: "service_call",
    autoInstrumentable: false,
  },

  // ── Databases ───────────────────────────────────────────────────────
  // MongoDB
  {
    regex: /from\s+["']mongodb["']|require\s*\(\s*["']mongodb["']\s*\)/g,
    provider: "mongodb",
    type: "sdk_import",
    autoInstrumentable: false,
  },
  {
    regex: /\.\s*(find|findOne|insertOne|insertMany|updateOne|updateMany|deleteOne|deleteMany|aggregate)\s*\(/g,
    provider: "mongodb",
    type: "service_call",
    autoInstrumentable: false,
  },
  // Supabase
  {
    regex: /from\s+["']@supabase\/supabase-js["']/g,
    provider: "supabase",
    type: "sdk_import",
    autoInstrumentable: false,
  },
  // Firebase
  {
    regex: /from\s+["']firebase-admin["']|from\s+["']firebase\/firestore["']/g,
    provider: "firebase",
    type: "sdk_import",
    autoInstrumentable: false,
  },

  // ── Search ──────────────────────────────────────────────────────────
  {
    regex: /from\s+["']@elastic\/elasticsearch["']/g,
    provider: "elasticsearch",
    type: "sdk_import",
    autoInstrumentable: false,
  },

  // ── Cache ───────────────────────────────────────────────────────────
  {
    regex: /from\s+["'](ioredis|redis)["']|require\s*\(\s*["'](ioredis|redis)["']\s*\)/g,
    provider: "redis",
    type: "sdk_import",
    autoInstrumentable: false,
  },

  // ── Vector databases ────────────────────────────────────────────────
  {
    regex: /from\s+["']@pinecone-database\/pinecone["']/g,
    provider: "pinecone",
    type: "sdk_import",
    autoInstrumentable: false,
  },
  {
    regex: /\.\s*(query|upsert)\s*\(/g,
    provider: "vector-db",
    type: "service_call",
    autoInstrumentable: false,
  },
  {
    regex: /from\s+["']weaviate-ts-client["']|from\s+["']weaviate-client["']/g,
    provider: "weaviate",
    type: "sdk_import",
    autoInstrumentable: false,
  },
  {
    regex: /from\s+["']chromadb["']/g,
    provider: "chromadb",
    type: "sdk_import",
    autoInstrumentable: false,
  },
  {
    regex: /from\s+["']@qdrant\/js-client-rest["']/g,
    provider: "qdrant",
    type: "sdk_import",
    autoInstrumentable: false,
  },

  // ── Messaging / Communications ──────────────────────────────────────
  {
    regex: /from\s+["']twilio["']|require\s*\(\s*["']twilio["']\s*\)/g,
    provider: "twilio",
    type: "sdk_import",
    autoInstrumentable: false,
  },
  {
    regex: /from\s+["']@sendgrid\/mail["']/g,
    provider: "sendgrid",
    type: "sdk_import",
    autoInstrumentable: false,
  },
  {
    regex: /from\s+["']resend["']/g,
    provider: "resend",
    type: "sdk_import",
    autoInstrumentable: false,
  },
  {
    regex: /from\s+["']@slack\/web-api["']/g,
    provider: "slack",
    type: "sdk_import",
    autoInstrumentable: false,
  },

  // ── Geo / Maps ──────────────────────────────────────────────────────
  {
    regex: /from\s+["']@googlemaps\/google-maps-services-js["']/g,
    provider: "google-maps",
    type: "sdk_import",
    autoInstrumentable: false,
  },

  // ── Web scraping / data ─────────────────────────────────────────────
  {
    regex: /from\s+["']@mendable\/firecrawl-js["']/g,
    provider: "firecrawl",
    type: "sdk_import",
    autoInstrumentable: false,
  },
  {
    regex: /from\s+["']@tavily\/core["']/g,
    provider: "tavily",
    type: "sdk_import",
    autoInstrumentable: false,
  },
];

/**
 * Scan a source string for cost-generating patterns.
 *
 * @param source - File contents to scan.
 * @param fileName - Path to include in results.
 * @returns Array of detected cost points.
 */
export function scanSource(source: string, fileName: string): CostPoint[] {
  const results: CostPoint[] = [];
  const seenLines = new Set<string>(); // "provider:line" dedup key

  for (const pattern of PATTERNS) {
    // Reset regex state for each file
    pattern.regex.lastIndex = 0;
    let match: RegExpExecArray | null;

    while ((match = pattern.regex.exec(source)) !== null) {
      const beforeMatch = source.slice(0, match.index);
      const lineNumber = beforeMatch.split("\n").length;

      // Deduplicate: same provider on same line
      const key = `${pattern.provider}:${lineNumber}`;
      if (seenLines.has(key)) continue;
      seenLines.add(key);

      results.push({
        file: fileName,
        line: lineNumber,
        provider: pattern.provider,
        type: pattern.type,
        pattern: match[0].trim(),
        autoInstrumentable: pattern.autoInstrumentable,
      });
    }
  }

  return results;
}

/**
 * Generate integration stub code from detected cost points.
 *
 * Produces a self-contained template showing:
 * 1. SDK initialisation
 * 2. Customer/project context
 * 3. Task wrapper with manual `recordCost` calls
 * 4. Summary of auto-instrumented providers
 */
export function generateStubs(points: CostPoint[], _scanDir: string): string {
  const auto = points.filter((p) => p.autoInstrumentable);
  const manual = points.filter((p) => !p.autoInstrumentable);

  const lines: string[] = [
    "// ============================================================",
    "// dexcost integration stubs",
    "// Generated by: dexcost scan --generate-stubs",
    "// ============================================================",
    "",
    "// --- Step 1: Initialize dexcost ---",
    'import { init, track } from "@dexcost/sdk";',
    "",
    'init({ apiKey: "dx_live_..." }); // or set DEXCOST_API_KEY env var',
    "",
    "// --- Step 2: Set customer context (in your request handler) ---",
    'import { setContext } from "@dexcost/sdk";',
    "",
    "setContext({",
    '  customerId: "your_customer_id",',
    '  projectId: "your_project_id",',
    "});",
    "",
    "// --- Step 3: Track tasks ---",
    'await track({ taskType: "your_task_type" }, async (task) => {',
    "  // Your agent code here...",
    "",
  ];

  if (manual.length > 0) {
    lines.push(
      "  // --- Manual cost tracking (for services not auto-instrumented) ---",
    );
    for (const p of manual) {
      lines.push(`  // ${p.file}:${p.line} \u2014 ${p.pattern}`);
      lines.push(
        `  task.recordCost("${p.provider}", 0.00); // TODO: set actual cost`,
      );
      lines.push("");
    }
  }

  lines.push("});");
  lines.push("");

  if (auto.length > 0) {
    lines.push("// --- Auto-instrumented (no code changes needed) ---");
    const providers = new Map<string, number>();
    for (const p of auto) {
      providers.set(p.provider, (providers.get(p.provider) ?? 0) + 1);
    }
    for (const [provider, count] of providers) {
      lines.push(
        `// \u2713 ${provider} (${count} call${count > 1 ? "s" : ""} detected)`,
      );
    }
  }

  return lines.join("\n") + "\n";
}
